### ISSUE TITLE
nodb/nouser warning message

### ISSUE THREAD

**Issue by jasimul** [pgbouncer/pgbouncer#768]
Pgbouncer log showing some unexpected warning message of previously configured database.
 
```
2022-10-15 05:29:35.129 UTC [25061] LOG C-0x5613cb53dae0: (nodb)/(nouser)@10.46.93.14:58446 closing because: no such database: postgres14poc (age=0s)
2022-10-15 05:29:35.129 UTC [25061] WARNING C-0x5613cb53dae0: (nodb)/(nouser)@10.46.93.14:58446 pooler error: no such database: postgres14poc
2022-10-15 05:29:35.129 UTC [25061] LOG C-0x5613cb53dae0: (nodb)/(nouser)@10.46.93.14:58446 login failed: db=postgres14poc user=mr3master
2022-10-15 05:29:35.129 UTC [25061] LOG C-0x5613cb53ef80: (nodb)/(nouser)@10.46.93.14:58448 closing because: no such database: postgres14poc (age=0s)
2022-10-15 05:29:35.129 UTC [25061] WARNING C-0x5613cb53ef80: (nodb)/(nouser)@10.46.93.14:58448 pooler error: no such database: postgres14poc
2022-10-15 05:29:35.129 UTC [25061] LOG C-0x5613cb53ef80: (nodb)/(nouser)@10.46.93.14:58448 login failed: db=postgres14poc user=mr3master
```
This postgres14poc  db i have configured earlier  and now its has different entry , not sure from where its coming.
Though pgbouncer admin console shows proper DB entry.

**Things i have tried :**
1. reload pgbouncer daemon
2. restart DB server 

But still its coming 
**pgbouncer details  :**  
psql (12.12 (Ubuntu 12.12-0ubuntu0.20.04.1), server 1.12.0/bouncer)

```
pgbouncer=# show pools;
 database  |   user    | cl_active | cl_waiting | sv_active | sv_idle | sv_used | sv_tested | sv_login | maxwait | maxwait_us |  pool_mode
-----------+-----------+-----------+------------+-----------+---------+---------+-----------+----------+---------+------------+-------------
 perfdev   | mr3master |        11 |          0 |         1 |       1 |       2 |         0 |        0 |       0 |          0 | transaction
 pgbouncer | pgbouncer |         1 |          0 |         0 |       0 |       0 |         0 |        0 |       0 |          0 | statement
```

**here is pgbouncer.ini**
```ini
;;;
;;; PgBouncer configuration file
;;;

;; database name = connect string
;;
;; connect string params:
;;   dbname= host= port= user= password= auth_user=
;;   client_encoding= datestyle= timezone=
;;   pool_size= reserve_pool= max_db_connections=
;;   pool_mode= connect_query= application_name=
[databases]
postgres=dbname=postgres host=xxx.cluster-xxx.us-west-2.rds.amazonaws.com port=5432 pool_size=800
perfdev=dbname=perfdev host=xxx-cluster.cluster-xxx.us-west-2.rds.amazonaws.com port=5432 pool_size=800


;; foodb over Unix socket
;foodb =

;; redirect bardb to bazdb on localhost
;bardb = host=localhost dbname=bazdb

;; access to dest database will go with single user
;forcedb = host=127.0.0.1 port=300 user=baz password=foo client_encoding=UNICODE datestyle=ISO connect_query='SELECT 1'

;; use custom pool sizes
;nondefaultdb = pool_size=50 reserve_pool=10

;; use auth_user with auth_query if user not present in auth_file
;; auth_user must exist in auth_file
; foodb = auth_user=bar

;; fallback connect string
;* = host=testserver

;; User-specific configuration
[users]

;user1 = pool_mode=transaction max_user_connections=10

;; Configuration section
[pgbouncer]

;;;
;;; Administrative settings
;;;

logfile = /var/log/postgresql/pgbouncer.log
;logfile = pgbouncer.log
pidfile = /var/run/postgresql/pgbouncer.pid

;;;
;;; Where to wait for clients
;;;

;; IP address or * which means all IPs
listen_addr = *
listen_port = 6432

;; Unix socket is also used for -R.
;; On Debian it should be /var/run/postgresql
;unix_socket_dir = /tmp
;unix_socket_mode = 0777
;unix_socket_group =
unix_socket_dir = /var/run/postgresql

;;;
;;; TLS settings for accepting clients
;;;

;; disable, allow, require, verify-ca, verify-full
;client_tls_sslmode = disable
;client_tls_sslmode = allow
;; Path to file that contains trusted CA certs
;client_tls_ca_file = <system default>

;; Private key and cert to present to clients.
;; Required for accepting TLS connections from clients.
;client_tls_key_file =
;client_tls_cert_file =
;;client_tls_key_file = /etc/pgbouncer/key_file.pem
;;client_tls_cert_file = /etc/pgbouncer/cert_file.pem
;; fast, normal, secure, legacy, <ciphersuite string>
;client_tls_ciphers = fast
;
;; all, secure, tlsv1.0, tlsv1.1, tlsv1.2, tlsv1.3
;client_tls_protocols = all

;; none, auto, legacy
;client_tls_dheparams = auto

;; none, auto, <curve name>
;client_tls_ecdhcurve = auto

;;;
;;; TLS settings for connecting to backend databases
;;;

;; disable, allow, require, verify-ca, verify-full
;server_tls_sslmode = disable

;; Path to that contains trusted CA certs
;server_tls_ca_file = <system default>

;; Private key and cert to present to backend.
;; Needed only if backend server require client cert.
;server_tls_key_file =
;server_tls_cert_file =

;; all, secure, tlsv1.0, tlsv1.1, tlsv1.2, tlsv1.3
;server_tls_protocols = all

;; fast, normal, secure, legacy, <ciphersuite string>
;server_tls_ciphers = fast

;;;
;;; Authentication settings
;;;

;; any, trust, plain, md5, cert, hba, pam
auth_type = md5
auth_file = /etc/pgbouncer/userlist.txt

;; Path to HBA-style auth config
;auth_hba_file =

;; Query to use to fetch password from database.  Result
;; must have 2 columns - username and password hash.
;auth_query = SELECT usename, passwd FROM pg_shadow WHERE usename=$1

;;;
;;; Users allowed into database 'pgbouncer'
;;;

;; comma-separated list of users who are allowed to change settings
;admin_users = postgres, migratinator_test_sarfaraj, mirthresults
admin_users = pgbouncer_admin
;; comma-separated list of users who are just allowed to use SHOW command
;stats_users = stats, root

;;;
;;; Pooler personality questions
;;;

;; When server connection is released back to pool:
;;   session      - after client disconnects (default)
;;   transaction  - after transaction finishes
;;   statement    - after statement finishes
;pool_mode = session




;; Query for cleaning connection immediately after releasing from
;; client.  No need to put ROLLBACK here, pgbouncer does not reuse
;; connections where transaction is left open.
;server_reset_query = DISCARD ALL

;; Whether server_reset_query should run in all pooling modes.  If it
;; is off, server_reset_query is used only for session-pooling.
;server_reset_query_always = 0

;; Comma-separated list of parameters to ignore when given in startup
;; packet.  Newer JDBC versions require the extra_float_digits here.
;ignore_startup_parameters = extra_float_digits

;; When taking idle server into use, this query is run first.
;server_check_query = select 1

;; If server was used more recently that this many seconds ago,
; skip the check query.  Value 0 may or may not run in immediately.
;server_check_delay = 30

;; Close servers in session pooling mode after a RECONNECT, RELOAD,
;; etc. when they are idle instead of at the end of the session.
;server_fast_close = 0

;; Use <appname - host> as application_name on server.
;application_name_add_host = 0

;; Period for updating aggregated stats.
;stats_period = 60

;;;
;;; Connection limits
;;;

;; Total number of clients that can connect
max_client_conn = 4000

;; Default pool size.  20 is good number when transaction pooling
;; is in use, in session pooling it needs to be the number of
;; max clients you want to handle at any moment
;default_pool_size = 100

;; Minimum number of server connections to keep in pool.
;min_pool_size = 0

; how many additional connection to allow in case of trouble
;reserve_pool_size = 0

;; If a clients needs to wait more than this many seconds, use reserve
;; pool.
;reserve_pool_timeout = 5

;; How many total connections to a single database to allow from all
;; pools
max_db_connections = 4000
max_user_connections = 4000
pool_mode = transaction
;pool_size = 800
;; If off, then server connections are reused in LIFO manner
;server_round_robin = 0

;;;
;;; Logging
;;;

;; Syslog settings
;syslog = 0
;syslog_facility = daemon
;syslog_ident = pgbouncer

;; log if client connects or server connection is made
;log_connections = 1

;; log if and why connection was closed
;log_disconnections = 1

;; log error messages pooler sends to clients
;log_pooler_errors = 1

;; write aggregated stats into log
;log_stats = 1

;; Logging verbosity.  Same as -v switch on command line.
;verbose = 0

;;;
;;; Timeouts
;;;

;; Close server connection if its been connected longer.
;server_lifetime = 3600

;; Close server connection if its not been used in this time.  Allows
;; to clean unnecessary connections from pool after peak.
;server_idle_timeout = 600

;; Cancel connection attempt if server does not answer takes longer.
;server_connect_timeout = 15

;; If server login failed (server_connect_timeout or auth failure)
;; then wait this many second.
;server_login_retry = 15

;; Dangerous.  Server connection is closed if query does not return in
;; this time.  Should be used to survive network problems, _not_ as
;; statement_timeout. (default: 0)
query_timeout = 300

;; Dangerous.  Client connection is closed if the query is not
;; assigned to a server in this time.  Should be used to limit the
;; number of queued queries in case of a database or network
;; failure. (default: 120)
query_wait_timeout = 300

;; Dangerous.  Client connection is closed if no activity in this
;; time.  Should be used to survive network problems. (default: 0)
;client_idle_timeout = 0

;; Disconnect clients who have not managed to log in after connecting
;; in this many seconds.
;client_login_timeout = 60

;; Clean automatically created database entries (via "*") if they stay
;; unused in this many seconds.
; autodb_idle_timeout = 3600

;; Close connections which are in "IDLE in transaction" state longer
;; than this many seconds.
;idle_transaction_timeout = 0

;; How long SUSPEND/-R waits for buffer flush before closing
;; connection.
;suspend_timeout = 10

;;;
;;; Low-level tuning options
;;;

;; buffer for streaming packets
;pkt_buf = 4096

;; man 2 listen
;listen_backlog = 128

;; Max number pkt_buf to process in one event loop.
;sbuf_loopcnt = 5

;; Maximum PostgreSQL protocol packet size.
;max_packet_size = 2147483647

;; Set SO_REUSEPORT socket option
;so_reuseport = 0

;; networking options, for info: man 7 tcp

;; Linux: Notify program about new connection only if there is also
;; data received.  (Seconds to wait.)  On Linux the default is 45, on
;; other OS'es 0.
;tcp_defer_accept = 0

;; In-kernel buffer size (Linux default: 4096)
;tcp_socket_buffer = 0

;; whether tcp keepalive should be turned on (0/1)
;tcp_keepalive = 1

;; The following options are Linux-specific.  They also require
;; tcp_keepalive=1.

;; Count of keepalive packets
;tcp_keepcnt = 0

;; How long the connection can be idle before sending keepalive
;; packets
;tcp_keepidle = 0

;; The time between individual keepalive probes
;tcp_keepintvl = 0

;; DNS lookup caching time
;dns_max_ttl = 15

;; DNS zone SOA lookup period
;dns_zone_check_period = 0

;; DNS negative result caching time
;dns_nxdomain_ttl = 15

;; Custom resolv.conf file, to set custom DNS servers or other options
;; (default: empty = use OS settings)
;resolv_conf = /etc/pgbouncer/resolv.conf

;;;
;;; Random stuff
;;;

;; Hackish security feature.  Helps against SQL injection: when PQexec
;; is disabled, multi-statement cannot be made.
;disable_pqexec = 0

;; Config file to use for next RELOAD/SIGHUP
;; By default contains config file from command line.
;conffile

;; Windows service name to register as.  job_name is alias for
;; service_name, used by some Skytools scripts.
;service_name = pgbouncer
;job_name = pgbouncer

;; Read additional config from other file
;%include /etc/pgbouncer/pgbouncer-other.ini
```

**Comment by jasimul** [comment]
any suggestion/input  would be a great help

**Comment by jasimul** [comment]
it showing the above mention warning for all the previously configure database, seems it some ehere caching it
`
2022-10-18 03:08:23.535 UTC [88220] WARNING C-0x5624b39e2e90: pgbouncerpoc/mr3master@10.46.89.12:60574 pooler error: pgbouncer cannot connect to server
2022-10-18 03:08:23.540 UTC [88220] LOG C-0x5624b39e2e90: (nodb)/(nouser)@10.46.80.132:40604 closing because: no such database: perfdev (age=0s)
2022-10-18 03:08:23.540 UTC [88220] WARNING C-0x5624b39e2e90: (nodb)/(nouser)@10.46.80.132:40604 pooler error: no such database: perfdev
2022-10-18 03:08:23.540 UTC [88220] LOG C-0x5624b39e2e90: (nodb)/(nouser)@10.46.80.132:40604 login failed: db=perfdev user=mr3master
2022-10-18 03:08:23.542 UTC [88220] LOG C-0x5624b39e2e90: pgbouncerpoc/mr3master@10.46.89.12:60576 login attempt: db=pgbouncerpoc user=mr3master tls=no
2022-10-18 03:08:23.542 UTC [88220] LOG C-0x5624b39e2e90: pgbouncerpoc/mr3master@10.46.89.12:60576 closing because: pgbouncer cannot connect to server (age=0s)
2022-10-18 03:08:23.542 UTC [88220] WARNING C-0x5624b39e2e90: pgbouncerpoc/mr3master@10.46.89.12:60576 pooler error: pgbouncer cannot connect to server
2022-10-18 03:08:23.542 UTC [88220] LOG C-0x5624b39e2e90: pgbouncerpoc/mr3master@10.46.89.12:60578 login attempt: db=pgbouncerpoc user=mr3master tls=no
2022-10-18 03:08:23.542 UTC [88220] LOG C-0x5624b39e2e90: pgbouncerpoc/mr3master@10.46.89.12:60578 closing because: pgbouncer cannot connect to server (age=0s)
2022-10-18 03:08:23.542 UTC [88220] WARNING C-0x5624b39e2e90: pgbouncerpoc/mr3master@10.46.89.12:60578 pooler error: pgbouncer cannot connect to server
2022-10-18 03:08:23.543 UTC [88220] LOG C-0x5624b39e2e90: (nodb)/(nouser)@10.46.80.132:40606 closing because: no such database: postgres14poc (age=0s)
2022-10-18 03:08:23.543 UTC [88220] WARNING C-0x5624b39e2e90: (nodb)/(nouser)@10.46.80.132:40606 pooler error: no such database: postgres14poc
2022-10-18 03:08:23.543 UTC [88220] LOG C-0x5624b39e2e90: (nodb)/(nouser)@10.46.80.132:40606 login failed: db=postgres14poc user=mr3master

`

**Comment by jasimul** [comment]
@goksgie  sorry for the inconvinence 
here is the content without commented line ,  kept it intentionally thinking i might be missing to set any properties properly
`
[databases]
pgbouncerpoc=dbname=pgbouncerpoc host=xxx.cluster-xxx.us-west-2.rds.amazonaws.com port=5432 pool_size=1000
[users]

;user1 = pool_mode=transaction max_user_connections=10

[pgbouncer]
;;;
;;; Administrative settings
;;;

logfile = /var/log/postgresql/pgbouncer.log
;logfile = pgbouncer.log
pidfile = /var/run/postgresql/pgbouncer.pid

listen_addr = *
listen_port = 6432
unix_socket_dir = /var/run/postgresql
auth_type = md5
auth_file = /etc/pgbouncer/userlist.txt
admin_users = pgbouncer_admin
max_client_conn = 4000
max_db_connections = 4000
max_user_connections = 4000
pool_mode = transaction
query_timeout = 300
query_wait_timeout = 300

`

**Comment by jasimul** [comment]
@goksgie  can you please check this once

**Comment by goksgie** [comment]
@jasimul , what does `show databases;` return on your setup?

Additionally, do you have active clients that try to use the old connection setup (i.e., old database settings)?

**Comment by jasimul** [comment]
@goksgie  show database ,show pools and all other adminstartation command shows proper  database configuration.
I can check the second point and keep you posted.
But what i am more interested , is there any way i can prevent those warning message in log ? its make my life difficult to see if any legitimate error is happening

**Comment by eulerto** [comment]
WARNING is not ERROR. From the database perspective, if a database does not exist, it is an error but for PgBouncer is just a warning. The reason is that client connects to PgBouncer so no error is reported. However, PgBouncer couldn't connect to database, hence, it reports a warning and expects that the admin fixes it soon.

**Comment by goksgie** [comment]
@eulerto , do you think it would make sense to disconnect clients that are not connected yet to the target database when an error response is returned from the target? This looks like a one-line change [here](https://github.com/pgbouncer/pgbouncer/blob/a25c87c138f3f38508e67ed6183b7db155635645/src/server.c#L72).

**Comment by jasimul** [comment]
@eulerto , i am not sure why its trying to connect a database  which have no active mention in  pgbouncer.ini.
for my test purpose i was switching database , the warning message started flowing  for previously configured whenever i switch to other DB

**Comment by Rammurthy5** [comment]
I get the same error. Has it been fixed ? @jasimul

**Comment by jasimul** [comment]
Some how its cache old db reference.
A fresh installation might help but  we didn't go ahead with pgbouncer due
to some other issues

**Comment by Rammurthy5** [comment]
May I ask what did you use then please ?

**Comment by jasimul** [comment]
We didn't use pgbouncer connections pooling,we switched to aws read replica

**Comment by JelteF** [comment]
I responded to @Rammurthy5 in the discussion he created here: https://github.com/pgbouncer/pgbouncer/discussions/915#discussioncomment-6667881

Based on the logs from @jasimul it looks like there was some client still trying to connect to `postgres14poc`, even though that database line was removed from the pgbouncer config. Probably some client was just not updated with the new database name and tried reconnecting.

The `C` in `C-0x5613cb53dae0` means that it's logged for a client. And the nodb/nouser means no pool was assigned yet when the warning happened. Which makes sense because the error says that no pool could be found in the configuration with given name.
